src/main.clj

```diff
@@ -1,2 +1,314 @@
+         (line-seq rdr)))))
+ "/Users/a/w/4clojure/aoc1-input.txt")
+
+; https://adventofcode.com/2021/day/1#part2
+((defn aoc1-part2
+   [file]
+   (with-open [rdr (clojure.java.io/reader file)]
+     (:count
+       (reduce
+         (fn [report depth]
+           (let [aggregated-depth (reduce (fn [x y] (+ x (Integer/parseInt y))) 0 (vec depth))]
+             (if (< (:prev-depth report) aggregated-depth)
+               ;  If depth has increased, increase the count.
+               (update (assoc report :prev-depth aggregated-depth) :count inc)
+               ;  Just update previous depth in other case.
+               (assoc report :prev-depth aggregated-depth))))
+         ; Set previous depth as maximum for initial value.
+         {:prev-depth Integer/MAX_VALUE :count 0}
+         ; Feed sequence of readings to reduce.
+         (partition 3 1 (line-seq rdr))))))
+ "/Users/a/w/4clojure/aoc1-input.txt")
+
+; https://adventofcode.com/2021/day/2#part1
+((defn aoc2-part1
+   [file]
+   (with-open [rdr (clojure.java.io/reader file)]
+     (let
+       [position
+        (reduce
+          (fn [position cmd]
+            (let [split-cmd (clojure.string/split cmd #" ")
+                  units (Integer/parseInt (nth split-cmd 1))
+                  command (nth split-cmd 0)]
+              (case command
+                "forward" (update position :horizontal + units)
+                "down" (update position :depth + units)
+                "up" (update position :depth - units))))
+          ; Starting position is 0
+          {:horizontal 0 :depth 0}
+          ; Feed sequence of readings to reduce.
+          (line-seq rdr))]
+       (* (:horizontal position) (:depth position)))))
+ "/Users/a/w/4clojure/aoc2-input.txt")
+
+
+; https://adventofcode.com/2021/day/2#part2
+((defn aoc2-part2
+   [file]
+   (with-open [rdr (clojure.java.io/reader file)]
+     (let
+       [position
+        (reduce
+          (fn [position cmd]
+            (let [split-cmd (clojure.string/split cmd #" ")
+                  units (Integer/parseInt (nth split-cmd 1))
+                  command (nth split-cmd 0)]
+              (case command
+                "forward" (as-> position p
+                                (update p :horizontal + units)
+                                (update p :depth + (* units (:aim p))))
+                "down" (update position :aim + units)
+                "up" (update position :aim - units))))
+          ; Starting position is 0
+          {:horizontal 0 :depth 0 :aim 0}
+          ; Feed sequence of readings to reduce.
+          (line-seq rdr))]
+       (* (:horizontal position) (:depth position)))))
+ "/Users/a/w/4clojure/aoc2-input.txt")
+
+; Mars Rover - https://github.com/priyaaank/MarsRover
+; QQQ: How to achieve extensibility in Clojure?
+; The way I see it, all inter-method calls/reduce are deeply coupled.
+; And if I were to look at this after a few months, I won't be able to make head/tail out of this :(
```

Comment:
I agree that as of now, this code (and even the AoC solutions) aren't easy to understand.

This is mostly because the problems haven't been broken down well. After reading the Mars Rover problem statement, one might intuitively come up with the following rough solution:

1. The input is a series of rover positions and associated commands.
2. We first find out how to compute a rover's final position given an initial position and its commands.
3. Now that we know how to do this, we apply this procedure to the input to get a series of final positions.

However, this structure is not reflected in the solution here. The solution is written as one big `reduce`, and most of the domain logic is mixed up in the body. The same applies to the AoC solutions.

Let's discuss more over a call.

Reply:
Summary of our discussion:

1. Break down the problem through domain modelling. The code should reflect the intuitive solution to the problem in our head.
2. Define data models and prefer maps over vectors for associative data. One doesn't require types to define a data model.
3. Loops, tail recursion and `reduce` are very low-level constructs. Prefer higher level functions like `map`, `filter`, `partition` or other higher-level domain functions.